### DISCUSSION TITLE
OC MultiTenant MultiTheme Resources and Static Files

### DISCUSSION THREAD

**Discussion by inzi** [OrchardCMS/OrchardCore#10578] [Q&A] [answered]
The most simple example of what I can't get working is as follows:

OC site, using Razor, multiple modules, multiple tenants, two themes. ThemeA, ThemeB. 

Both themes have site wide styles, images, and scripts. 

When the user logs in, they are at the InfoModule.  The info module has a generic class of object. 

Each theme styles this object differently. 

In the object, an IMG tag has a source that exists in both themes, each with their own version. A logo, for example.

If ThemeA is active, the relative path to the image is /ThemeA/images/logo.png.  

If ThemeB is active, the relative path to the image is /ThemeB/images/logo.png.

How can I set the proper relative path in the module to the logo, without having to specify the theme's path? 

Resource manifests require uniqueness, based on what I read, so I cannot have the module find the static file that way. 

How do i do I make a static file reference correctly?  Is there a way to automatically prepend the IMG's SRC with the the theme name? Something like src="~@OrchardCore.Resources.ThemeName/images/logo.png"? 

I am unclear how to accomplish this? An example to look at would be very helpful.

Thanks!

**Comment by deanmarcussen** [comment, marked as answer]
What I would suggest (without seeing your codebase)

Modules are designed to be themable, by a theme.

So the theme, when it is styling the `InfoModule` should be the one providing it's path to the image. 

The info module might know that an image should exist (but I would say it probably shouldn't)

I'm not sure if you've done this with shapes, or razor pages. 

But in my opinion the thing that is rendering the InfoModule should be in the theme, so the theme can style it.

Or you need some kind of provider, that the theme(s) implement to provide a img url. 

Hopefully that's some ideas to get you going again.

**Reply by inzi** [reply]
If, in my themes, their assets reside in wwwroot/img, wwwroot/css, wwwroot/js...

In the theme's startup - can I do something like this? 

            app.MapWhen(
                context => context.Request.Path.Value.StartsWith("/css", StringComparison.InvariantCultureIgnoreCase) ||
                        context.Request.Path.Value.StartsWith("/img", StringComparison.InvariantCultureIgnoreCase) ||
                        context.Request.Path.Value.StartsWith("/js", StringComparison.InvariantCultureIgnoreCase),
                config => config.UseStaticFiles()
            );

Should I do that in all the modules instead? 

I know the ordering is critical, I'm not sure where to put it so it applies everywhere.

**Reply by inzi** [reply]
I'm not doing shapes yet, just razor. I think I understand what you're saying. I'm still learning some basic concepts on implementing OC.

Perhaps I'm not grasping how themes are used. I am believing that:

The theme provides the layout, libraries, and static files.

in the module, at the controller's method's view, I only render the body.

The body for that controller's method is rendered inside the layout page in the theme.

In effect, the "page" that is served to the client is made up the Layout file in the theme, and then the view from the model is injected when the Layout invokes RenderBodyAsync()? Is this correct?
